zk/smt/unwind_smt.go

```diff
@@ -39,173 +42,240 @@ func UnwindZkSMT(ctx context.Context, logPrefix string, from, to uint64, tx kv.R
-		eridb.OpenBatch(ctx.Done())
+		quit := make(chan struct{})
+		eridb.OpenBatch(quit)
 	}
 
-	ac, err := tx.CursorDupSort(kv.AccountChangeSet)
-	if err != nil {
-		return trie.EmptyRoot, err
-	}
-	defer ac.Close()
-
-	sc, err := tx.CursorDupSort(kv.StorageChangeSet)
-	if err != nil {
-		return trie.EmptyRoot, err
+	changesGetter := NewChangesGetter(tx)
+	if err := changesGetter.openChangesGetter(from); err != nil {
+		return trie.EmptyRoot, fmt.Errorf("OpenChangesGetter: %w", err)
 	}
-	defer sc.Close()
-
-	currentPsr := state.NewPlainStateReader(tx)
+	defer changesGetter.closeChangesGetter()
 
 	total := uint64(math.Abs(float64(from) - float64(to) + 1))
-	printerStopped := false
 	progressChan, stopPrinter := zk.ProgressPrinter(fmt.Sprintf("[%s] Progress unwinding", logPrefix), total, quiet)
-	defer func() {
-		if !printerStopped {
-			stopPrinter()
-		}
-	}()
+	defer stopPrinter()
 
 	// walk backwards through the blocks, applying state changes, and deletes
 	// PlainState contains data AT the block
 	// History tables contain data BEFORE the block - so need a +1 offset
-	accChanges := make(map[common.Address]*accounts.Account)
-	codeChanges := make(map[common.Address]string)
-	storageChanges := make(map[common.Address]map[string]string)
-
-	addDeletedAcc := func(addr common.Address) {
-		deletedAcc := new(accounts.Account)
-		deletedAcc.Balance = *uint256.NewInt(0)
-		deletedAcc.Nonce = 0
-		accChanges[addr] = deletedAcc
-	}
-
-	psr := state.NewPlainState(tx, from, systemcontracts.SystemContractCodeLookup["Hermez"])
-	defer psr.Close()
-
 	for i := from; i >= to+1; i-- {
 		select {
 		case <-ctx.Done():
-			return trie.EmptyRoot, fmt.Errorf("[%s] Context done", logPrefix)
+			return trie.EmptyRoot, fmt.Errorf("context done")
 		default:
 		}
 
-		psr.SetBlockNr(i)
+		if err := changesGetter.getChangesForBlock(i); err != nil {
+			return trie.EmptyRoot, fmt.Errorf("getChangesForBlock: %w", err)
+		}
 
-		dupSortKey := dbutils.EncodeBlockNumber(i)
+		progressChan <- 1
+	}
 
-		// collect changes to accounts and code
-		for _, v, err2 := ac.SeekExact(dupSortKey); err2 == nil && v != nil; _, v, err2 = ac.NextDup() {
+	stopPrinter()
 
-			addr := common.BytesToAddress(v[:length.Addr])
+	if _, _, err := dbSmt.SetStorage(ctx, logPrefix, changesGetter.accChanges, changesGetter.codeChanges, changesGetter.storageChanges); err != nil {
+		return trie.EmptyRoot, err
+	}
 
-			// if the account was created in this changeset we should delete it
-			if len(v[length.Addr:]) == 0 {
-				codeChanges[addr] = ""
-				addDeletedAcc(addr)
-				continue
-			}
+	lr := dbSmt.LastRoot()
 
-			oldAcc, err := psr.ReadAccountData(addr)
-			if err != nil {
-				return trie.EmptyRoot, err
-			}
+	hash := common.BigToHash(lr)
+	if checkRoot && hash != *expectedRootHash {
+		log.Error("failed to verify hash")
+		return trie.EmptyRoot, fmt.Errorf("wrong trie root: %x, expected (from header): %x", hash, expectedRootHash)
```

Comment:
Is it emptyerror or just erong root ? Does the error name matter here ?

Reply:
It is wrong root but now that you mention it, this is repeating code. We already have a function to verify the root and can use it here instead of that. I'll make the change now.

Reply:
But I already have another branch with changes around that. I'll make the change there. It is branched out of this one, so we need to first merge this, before we can merge the other.